#Gatherplot IPython Demo Using Airline Data

This notebook shows how we can use the gatherplot with ipython 

## Data Source

* 2008 Airline Data from http://stat-computing.org/dataexpo/2009/   

* Gatherplot [Gatherplot.org](http://gatherplot.org/)

* IPython [ipython.org](http://ipython.org/)

In [1]:
import pandas as pd
import numpy as np

In [2]:
alldata = pd.read_csv('2008.csv')
alldata.shape

(7009728, 29)

## Data Cleaning

Let's remove some of destinations/origin/carrier.

First We will see histogram of data to get a few major destinations.  

Among all the carrier, we will only select WN, AA, and OO. 

Among all the destination and origin, we will select only 'DFW','ORD','ATL','LAX','PHX'

This filtering reduced the number of records from 7009728 to 67873.

In [3]:
filtered_data = alldata[alldata['UniqueCarrier'].isin(['WN','AA','OO'])]

filtered_data = filtered_data[filtered_data['Origin'].isin(['DFW','ORD','ATL','LAX','PHX'])]
filtered_data['Origin'].value_counts()

filtered_data = filtered_data[filtered_data['Dest'].isin(['DFW','ORD','ATL','LAX','PHX'])]
filtered_data['Dest'].value_counts()

filtered_30k = filtered_data.loc[np.random.choice(filtered_data.index, 30000, replace=False)]

filtered_30k.shape

(30000, 29)

## Gatherplot upload

This will upload the data to gatherplot and display as an iframe.  

In [4]:
import dropbox
import re
import time
from firebase import firebase
from IPython.display import IFrame

def publish_gatherplot(p, dataname):
    
    filename = str(time.time())+'.csv'
    p.index.name = 'indexPandas'
    p.to_csv(filename)
    access_token = 'vwezSddRAD0AAAAAAAA0K2kAPqcpKnsShlfAPyIxSAYG0l2GxKfx5vtEkSyCeNOk'
    client = dropbox.client.DropboxClient(access_token)
    f = open(filename,'rb')
    response = client.put_file('./gatherplot/' + filename,f)
    print "uploaded:", response
    
    response = client.share('./gatherplot/' + filename, short_url=False)
    print 'Shared:', response
    url = response['url']
    match = re.search(r'https://www\.dropbox\.com/s/(.+)\?dl=0', url)
    if match:
#     print match.group(1)
        download_url = 'https://dl.dropboxusercontent.com/s/' + match.group(1)
        print download_url
        
    aFirebase = firebase.FirebaseApplication('https://gatherplot-dev.firebaseio.com',None)
    
    new_csv = {'name':dataname, 
           'uploader':"google:110953151430048855242", 
           'uploaderName': "Deok Gun Park",
           'url': download_url } 

    result = aFirebase.post('/csv',new_csv)
    print result
    

    return 'https://gatherplot-dev.firebaseapp.com/#/load/' + result['name']

In [5]:
# data_30k = alldata[:30000]
# data_10k = alldata.loc[np.random.choice(alldata.index, 10000, replace=False)]

url = publish_gatherplot(filtered_30k, "Cody")
IFrame(url, 900,900)

uploaded: {u'icon': u'page_white', u'bytes': 3583380, u'thumb_exists': False, u'rev': u'2638a00af1980', u'modified': u'Wed, 25 Nov 2015 17:39:24 +0000', u'shareable': False, u'client_mtime': u'Wed, 25 Nov 2015 17:39:24 +0000', u'path': u'/gatherplot/1448473158.97.csv', u'is_dir': False, u'size': u'3.4 MB', u'root': u'dropbox', u'mime_type': u'text/csv', u'revision': 156554}
Shared: {u'url': u'https://www.dropbox.com/s/ep06nif62nrgi3g/1448473158.97.csv?dl=0', u'expires': u'Tue, 01 Jan 2030 00:00:00 +0000', u'visibility': u'PUBLIC'}
https://dl.dropboxusercontent.com/s/ep06nif62nrgi3g/1448473158.97.csv
{u'name': u'-K3zicA9Yqan93kI15lC'}


In [35]:
data_30k.head()

Year 
 Month 
 DayofMonth 
 DayOfWeek 
 DepTime 
 CRSDepTime 
 ArrTime 
 CRSArrTime 
 UniqueCarrier 
 FlightNum 
 ... 
 TaxiIn 
 TaxiOut 
 Cancelled 
 CancellationCode 
 Diverted 
 CarrierDelay 
 WeatherDelay 
 NASDelay 
 SecurityDelay 
 LateAircraftDelay 
 
 
 indexPandas 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 0 
 2008 
 1 
 3 
 4 
 2003 
 1955 
 2211 
 2225 
 WN 
 335 
 ... 
 4 
 8 
 0 
 NaN 
 0 
 NaN 
 NaN 
 NaN 
 NaN 
 NaN 
 
 
 1 
 2008 
 1 
 3 
 4 
 754 
 735 
 1002 
 1000 
 WN 
 3231 
 ... 
 5 
 10 
 0 
 NaN 
 0 
 NaN 
 NaN 
 NaN 
 NaN 
 NaN 
 
 
 2 
 2008 
 1 
 3 
 4 
 628 
 620 
 804 
 750 
 WN 
 448 
 ... 
 3 
 17 
 0 
 NaN 
 0 
 NaN 
 NaN 
 NaN 
 NaN 
 NaN 
 
 
 3 
 2008 
 1 
 3 
 4 
 926 
 930 
 1054 
 1100 
 WN 
 1746 
 ... 
 3 
 7 
 0 
 NaN 
 0 
 NaN 
 NaN 
 NaN 
 NaN 
 NaN 
 
 
 4 
 2008 
 1 
 3 
 4 
 1829 
 1755 
 1959 
 1925 
 WN 
 3920 
 ... 
 3 
 10 
 0 
 NaN 
 0 
 2 
 0 
 0 
 0 
 32 
 
 
 
 5 rows × 29 columns

In [13]:
# data_30k = alldata[:30000]
data_3k = alldata.loc[np.random.choice(alldata.index, 3000, replace=False)]

url = publish_gatherplot(data_10k, "Temp Airline On-time Performance random 3K data")
IFrame(url, 900,900)

uploaded: {u'icon': u'page_white', u'bytes': 1177516, u'thumb_exists': False, u'rev': u'250f900af1980', u'modified': u'Tue, 24 Mar 2015 14:36:25 +0000', u'shareable': False, u'client_mtime': u'Tue, 24 Mar 2015 14:36:25 +0000', u'path': u'/gatherplot/1427207787.4.csv', u'is_dir': False, u'size': u'1.1 MB', u'root': u'dropbox', u'mime_type': u'text/csv', u'revision': 151801}
Shared: {u'url': u'https://www.dropbox.com/s/vbgu0skq27k96od/1427207787.4.csv?dl=0', u'expires': u'Tue, 01 Jan 2030 00:00:00 +0000', u'visibility': u'PUBLIC'}
https://dl.dropboxusercontent.com/s/vbgu0skq27k96od/1427207787.4.csv
{u'name': u'-JlBCdFsyrTHY9RlpydK'}


In [14]:
# data_30k = alldata[:30000]
data_10k = alldata.loc[np.random.choice(alldata.index, 10000, replace=False)]

url = publish_gatherplot(data_10k, "Temp Airline On-time Performance random 10K data")
IFrame(url, 900,900)

uploaded: {u'icon': u'page_white', u'bytes': 1180442, u'thumb_exists': False, u'rev': u'250fa00af1980', u'modified': u'Tue, 24 Mar 2015 15:08:08 +0000', u'shareable': False, u'client_mtime': u'Tue, 24 Mar 2015 15:08:08 +0000', u'path': u'/gatherplot/1427209690.57.csv', u'is_dir': False, u'size': u'1.1 MB', u'root': u'dropbox', u'mime_type': u'text/csv', u'revision': 151802}
Shared: {u'url': u'https://www.dropbox.com/s/ss1qoq49tqd1eg1/1427209690.57.csv?dl=0', u'expires': u'Tue, 01 Jan 2030 00:00:00 +0000', u'visibility': u'PUBLIC'}
https://dl.dropboxusercontent.com/s/ss1qoq49tqd1eg1/1427209690.57.csv
{u'name': u'-JlBJtgufYoXKqgV7AEr'}
